src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java

```diff
@@ -138,6 +145,82 @@ public void updateWithReaction(Post post, Reaction reaction) {
         postRepository.save(post);
     }
 
+    /**
+     * @param courseId          id of the course the fetch posts for
+     * @param courseWideContext course-wide context for which the posts should be fetched
+     * @param exerciseId        id of the exercise for which the posts should be fetched
+     * @param lectureId         id of the lecture for which the posts should be fetched
+     * @return list of posts that match the given context
+     */
+    public List<Post> getPostsInCourse(Long courseId, CourseWideContext courseWideContext, Long exerciseId, Long lectureId) {
+        // no filter -> get all posts in course
+        if (courseWideContext == null && exerciseId == null && lectureId == null) {
+            return this.getAllCoursePosts(courseId);
+        }
+        // filter by course-wide context
+        else if (courseWideContext != null && exerciseId == null && lectureId == null) {
+            return this.getAllPostsByCourseWideContext(courseId, courseWideContext);
+        }
+        // filter by exercise
+        else if (courseWideContext == null && exerciseId != null && lectureId == null) {
+            return this.getAllExercisePosts(courseId, exerciseId);
+        }
+        // filter by lecture
+        else if (courseWideContext == null && exerciseId == null && lectureId != null) {
+            return this.getAllLecturePosts(courseId, lectureId);
+        }
+        else {
+            throw new BadRequestAlertException("A new post cannot be associated with more than one context", METIS_POST_ENTITY_NAME, "ambiguousContext");
+        }
+    }
+
+    /**
+     * Checks course, user and post validity,
+     * retrieves all posts for a course by its id
+     * and ensures that sensitive information is filtered out
+     *
+     * @param courseId id of the course the post belongs to
+     * @return list of posts that belong to the course
+     */
+    public List<Post> getAllCoursePosts(Long courseId) {
+        final User user = userRepository.getUserWithGroupsAndAuthorities();
+
+        // checks
+        Course course = preCheckUserAndCourse(user, courseId);
```

Comment:
To get this right: You mean duplicated in the sense that they are called in each of the methods not that they are called several times with the same input?
I am thinking about refactoring the resources in hand with the services anyways after I have spoken with the resource refactoring expert @PatZintel :D So it would rather stick to the current version and try to find a better solution once I updated the resources and service so that they adhere out guidelines.

Reply:
Yes, duplicated in the sense that the same line is written multiple times. Functionality-wise it makes sense to me. But if this is to be refactored anyways and out of scope for this PR, thats also fine with me

Reply:
Yes, duplicated in the sense that the same line is written multiple times. Functionality-wise it makes sense to me. But if this is to be refactored anyways and out of scope for this PR, thats also fine with me